// @hotloop/hotloop-sdk/dist/HotLoop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HotLoopSdkFactory = exports.HotLoop = void 0;
var TokenProvider_1 = require("./TokenProvider");
var axiosRetry = require('axios-retry');
var axios_1 = require("axios");
var Correlation_1 = require("./Correlation");
var Enums_1 = require("./Enums");
/**
 * Creates hydrated HotLoopSdk instances
 */
var HotLoopSdkFactory = /** @class */ (function () {
    function HotLoopSdkFactory() {
    }
    /**
     * Get a hydrated HotLoopSdk instance
     * @param key The HotLoop API key
     * @param opts The SDK options
     */
    HotLoopSdkFactory.getInstance = function (key, opts) {
        var url = 'https://europe-west3-hotloop-289416.cloudfunctions.net/';
        var tokenProvider = new TokenProvider_1.GcpTokenProvider(key);
        return new HotLoop(url, tokenProvider, opts);
    };
    return HotLoopSdkFactory;
}());
exports.HotLoopSdkFactory = HotLoopSdkFactory;
/**
 * The HotLoop SDK implementation
 */
var HotLoop = /** @class */ (function () {
    /**
     * @constructor
     * @param url The base URL for the HotLoop API
     * @param tokenProvider A mechanism for fetching tokens for the request
     * @param opts The options required to configure the SDK
     */
    function HotLoop(url, tokenProvider, opts) {
        var _this = this;
        this.tokenProvider = tokenProvider;
        var config = {
            baseURL: url,
            timeout: opts.timeout || 5000,
            headers: {
                'Accept': 'application/json; charset=utf-8',
                'User-Agent': opts.userAgent || 'hotloop-sdk'
            }
        };
        this.axios = axios_1.default.create(config);
        this.axios.interceptors.request.use(function (config) {
            if (!config.baseURL)
                throw new Error('No baseUrl in request config');
            if (!config.url)
                throw new Error('No URL in request config');
            // config.url is not the full URL. Temporary solution, see https://github.com/axios/axios/pull/2555
            var absoluteUrl = config.baseURL.replace(/\/+$/, '') + axios_1.default.getUri(config);
            return _this.tokenProvider.getBearerToken(absoluteUrl)
                .then(function (token) {
                config.headers['Authorization'] = "Bearer " + token;
                return config;
            });
        });
        var retryConfig = {
            retries: opts.retries || 3,
            retryDelay: function (retryCount) { return retryCount * (opts.retryDelay || 1000); }
        };
        axiosRetry(this.axios, retryConfig);
    }
    /**
     * @inheritDoc
     */
    HotLoop.prototype.syncDeployment = function (options) {
        return this.axios.post(Enums_1.Endpoint.SYNC_DEPLOYMENT, options)
            .then(Correlation_1.correlationIdFromResponse);
    };
    /**
     * @inheritDoc
     */
    HotLoop.prototype.syncCoverage = function (options) {
        return this.axios.post(Enums_1.Endpoint.SYNC_COVERAGE, options)
            .then(Correlation_1.correlationIdFromResponse);
    };
    return HotLoop;
}());
exports.HotLoop = HotLoop;
